server/aesdsocket.c

```diff
@@ -58,23 +132,222 @@ static void signal_handler (int signo)
     exit (EXIT_SUCCESS);
 }
 
-int main(int argc, char *argv[])
+
+void receive_send_data(void* thread_param)
+{
+
+	thread_data_t *thread_func_args = (thread_data_t*)thread_param;
+
+	int received_bytes, write_bytes, read_bytes, send_bytes;
+	int buf_loc = 0;
+	int realloc_count = 1;
+
+        thread_func_args->buffer = (char *)malloc(MAXDATASIZE*sizeof(char));
+        if(thread_func_args->buffer == NULL)
+        {
+                syslog(LOG_ERR, "thread_func_args malloc error\n");
+                close_all();
+        }
+
+        thread_func_args->transmit_buffer = (char *)malloc(MAXDATASIZE*sizeof(char));
+        if(thread_func_args->transmit_buffer == NULL)
+        {
+                syslog(LOG_ERR, "thread_func_args malloc error\n");
+                close_all();
+        }
+
+
+
+	//data receive from client
+	do
+	{
+                received_bytes = recv(thread_func_args->client_fd, thread_func_args->buffer + buf_loc, MAXDATASIZE, 0); //sizeof(buffer)
+                if(received_bytes <= 0)
+                {
+                        syslog(LOG_ERR,"Error receiving\n");
+                }
+
+                buf_loc += received_bytes;
+
+                // increase buffer size dynamically if required
+                realloc_count += 1;
+                thread_func_args->buffer = (char *)realloc(thread_func_args->buffer, realloc_count*MAXDATASIZE*(sizeof(char)));
+                if(thread_func_args->buffer == NULL)
+                {
+                        close(thread_func_args->client_fd);
+                        close_all();
+                        syslog(LOG_ERR, "Realloc error\n");
+			exit(EXIT_FAILURE);
+                }
+
+
+	}while(strchr(thread_func_args->buffer, '\n') == NULL);
+	received_bytes = buf_loc;
+	thread_func_args->buffer[received_bytes] = '\0';
+	//total_bytes += received_bytes; //keep track of total bytes
+
+	int rc;
+	rc = pthread_mutex_lock(&main_mutex); //lock mutex
+	if(rc != 0)
+	{
+		syslog(LOG_ERR,"Mutex lock failure\n");
+		close_all();
+		exit (EXIT_FAILURE);
```

Comment:
@Sharath-Jonnala I'm not sure you need to exit the entire process in this case.  What would happen if you changed one of these error paths to use `return` after locking your mutex?